examples/run_glue_tpu.py

```diff
@@ -0,0 +1,617 @@
+    scheduler = get_linear_schedule_with_warmup(
+        optimizer, num_warmup_steps=args.warmup_steps, num_training_steps=t_total,
+    )
+
+    # Train!
+    logger.info("***** Running training *****")
+    logger.info("  Num examples = %d", len(dataloader) * args.train_batch_size)
+    logger.info("  Num Epochs = %d", args.num_train_epochs)
+    logger.info("  Instantaneous batch size per TPU core = %d", args.train_batch_size)
+    logger.info(
+        "  Total train batch size (w. parallel, distributed & accumulation) = %d",
+        (args.train_batch_size * args.gradient_accumulation_steps * xm.xrt_world_size()),
+    )
+    logger.info("  Gradient Accumulation steps = %d", args.gradient_accumulation_steps)
+    logger.info("  Total optimization steps = %d", t_total)
+
+    global_step = 0
+    loss = None
+    model.zero_grad()
+    train_iterator = trange(int(args.num_train_epochs), desc="Epoch", disable=disable_logging)
+    set_seed(args.seed)  # Added here for reproductibility (even between python 2 and 3)
+    for epoch in train_iterator:
+        # Get TPU parallel loader which sends data to TPU in background.
+        train_dataloader = pl.ParallelLoader(dataloader, [args.device]).per_device_loader(args.device)
+        epoch_iterator = tqdm(train_dataloader, desc="Iteration", total=len(dataloader), disable=disable_logging)
+        for step, batch in enumerate(epoch_iterator):
+
+            # Save model checkpoint.
+            if args.save_steps > 0 and global_step % args.save_steps == 0:
+                output_dir = os.path.join(args.output_dir, "checkpoint-{}".format(global_step))
+                logger.info("Saving model checkpoint to %s", output_dir)
+
+                if xm.is_master_ordinal():
+                    if not os.path.exists(output_dir):
+                        os.makedirs(output_dir)
+                    torch.save(args, os.path.join(output_dir, "training_args.bin"))
+
+                # Barrier to wait for saving checkpoint.
+                xm.rendezvous("mid_training_checkpoint")
+                # model.save_pretrained needs to be called by all ordinals
+                model.save_pretrained(output_dir)
+
+            model.train()
+            inputs = {"input_ids": batch[0], "attention_mask": batch[1], "labels": batch[3]}
+            if args.model_type != "distilbert":
+                # XLM, DistilBERT and RoBERTa don't use segment_ids
+                inputs["token_type_ids"] = batch[2] if args.model_type in ["bert", "xlnet"] else None
+            outputs = model(**inputs)
+            loss = outputs[0]  # model outputs are always tuple in transformers (see doc)
+
+            if args.gradient_accumulation_steps > 1:
+                loss = loss / args.gradient_accumulation_steps
+
+            loss.backward()
+
+            if (step + 1) % args.gradient_accumulation_steps == 0:
+                torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_grad_norm)
```

Comment:
I was under the impression that this line was very slow on TPU?

Reply:
https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md

Reply:
I think you're referring to the comment made here: https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md I've sent our this [PR](https://github.com/pytorch/xla/pull/1898) to clarify.

tl;dr: we've patched `torch.nn.utils.clip_grad_norm_` so that it's not slow.

Reply:
Looks like our comments crossed haha

Reply:
Ah good to know.